examples/doremi/doremi/dataloader.py

```diff
@@ -0,0 +1,363 @@
+
+        num_samples_per_dp_rank = self.batch_size * self.num_microbatches
+        dp_start_idx = self.rank * num_samples_per_dp_rank
+        dp_end_idx = dp_start_idx + num_samples_per_dp_rank
+
+        if dp_end_idx > len(self.batch):
+            raise StopIteration
+
+        dp_batch = self.batch[dp_start_idx:dp_end_idx]
+
+        assert len(dp_batch) == self.num_microbatches * self.batch_size
+
+        microbatch_start_idx = self.microbatch_idx * self.batch_size
+        microbatch_end_idx = microbatch_start_idx + self.batch_size
+
+        if microbatch_end_idx > len(dp_batch):
+            raise StopIteration
+
+        microbatch_idxs = dp_batch[microbatch_start_idx:microbatch_end_idx]
+
+        if self.microbatch_idx == self.num_microbatches - 1:
+            self.microbatch_idx = 0
+        else:
+            self.microbatch_idx += 1
+
+        return microbatch_idxs
+
+    def _round_up_domain_batch_sizes(self, domain_batch_size: List[int], target_total_size: int) -> List[int]:
+        """
+        NOTE: Make sum(domain_batch_sizes) == batch_size
+        """
+        total_batch_size = sum(domain_batch_size)
+        while total_batch_size != target_total_size:
+            diff = target_total_size - total_batch_size
+
+            # NOTE: Randomly select a domain to increase/decrase a sample
+            # to match the target_total_size
+            eligible_indices = torch.nonzero(torch.tensor(domain_batch_size) > 1).view(-1)
+            random_index = torch.randint(
+                low=0, high=len(eligible_indices), size=(1,), generator=self.generator, device="cpu"
+            ).item()
+            selected_domain = eligible_indices[random_index].item()
+
+            if diff > 0:
+                domain_batch_size[selected_domain] += 1
+            elif diff < 0 and domain_batch_size[selected_domain] > 0:
+                domain_batch_size[selected_domain] -= 1
+
+            total_batch_size = sum(domain_batch_size)
+
+        return domain_batch_size
+
+    def reset(self):
+        """Reset the state of the sampler for a new epoch."""
+        self.microbatch_idx = 0
+        self.domain_counters = [0 for _ in self.datasets]
+        self.total_samples_yielded = 0
+        self.out_of_samples = False
+
+        domain_indices = []
+        for i, dataset in enumerate(self.datasets):
+            local_indices = torch.arange(0, len(dataset), device="cpu").tolist()
+
+            # NOTE: align the indices across the combined dataset
+            global_indices = local_indices + self.offsets[i]
+            domain_indices.append(global_indices)
+
+        self.num_samples_per_global_step = self.batch_size * self.num_microbatches * self.num_replicas
+        self.domain_indices = domain_indices
+        self.expected_total_samples = sum([len(d) for d in domain_indices])
+
+
+def get_datasets(paths):
+    datasets = []
+    for path in tqdm(paths, desc="Loading dataset from disk"):
+        d = load_from_disk(path)
```

Comment:
Yes. But at the time I ran DoReMi, there wasn't a full version of the DoReMi dataset on the hub, so I downloaded and do stuff locally. We could share the dataset we used, so users wanting to reproduce this can use that dataset. But if they want to use other datasets, they could modify it themselves..